docs/blueprints/forgotten-deletes.md

```diff
@@ -0,0 +1,124 @@
+
+## Design
+
+The solution to this problem is twofold, as there are two ways in which pieces are deleted: delete pieces requests and garbage collection.
+
+### Garbage collection
+
+Garbage collection works by observing the metainfo loop. This collects all the piece IDs which a node should have according to the pointer DB at that time. The satellite then sends these piece IDs to the storage node, indicating that it should keep these pieces and all pieces which were created after the creation of the bloom filter and delete the rest.
+
+Instead of proving the pieces were deleted, we will remove the possibility of these pieces being restored.
+To acheive this we will simply run garbage collection from a DB snapshot at a point in time beyond which we will never restore. That is, we must decide what is the furthest possible point the past we may restore to and only run garbage collection on a snapshot at or before this point. If this is guaranteed, no pieces deleted by garbage collection should ever be reverted by a DB restoration.
+
+### Delete pieces requests
+
+Delete pieces requests are issued when a user deletes an object. The metainfo service deletes the object's pointers and then sends delete pieces requests to each of the reliable nodes which held pieces. The rest of the nodes' pieces will be eventually deleted by garbage collection. 
+
+To prove to the satellite that a piece was deleted legitimately, we will have it sign the delete pieces requests. The node will present these to the satellite in the event of an audit for a deleted piece.
+
+```
+message DeletePiecesRequest {
+    repeated bytes piece_ids = 1 [(gogoproto.customtype) = "PieceID", (gogoproto.nullable) = false];
+    bytes satellite_signature = 2;
+}
+```
+
+The storage nodes will serialize and save these signed messages into a file store.
+
+#### File store
+
+// TODO: directory names
+Within the storage node's 'storage' directory, we will add a new directory called 'deleted-pieces-proof'. Within this directory we will create a subdirectory for each satellite, and within each of these we will add two more directories: 'pieces' and 'signed-requests'.
+
+##### Creating Proofs
+
+The DeletePieces method on the piecestore endpoint is where the the delete pieces requests are sent. From here, the piece IDs within the message are enqueued for the piece deleter job to delete. Before the pieces are enqueued, we will write the serialized message into a file in the 'signed-requests' directory of the respective satellite, skipping the first byte which we will reserve. Then we pass this filepath along with the piece IDs into the Enqueue method. As pieces in the queue are deleted, we create a new path for them in the 'pieces' directory. At this path, we write a text file containing the name of the signed delete request file which contains this piece. Then, we open the signed delete request file and increment the first byte.
```

Comment:
I have a bunch of different questions about this section:
1. Is there a separate file for each delete pieces request? Or are multiple delete pieces requests appended to a single file? (some aspects of the following questions might not be relevant if there is only one request per file, but other aspects might still be relevant)
2. What is the first byte needed for? If it is only one byte, the maximum we could increment it to would be 255. What if a storagenode receives more piece deletion requests?
3. What if we added the message to a file in the `signed-requests` directory, then just symlinked this file from the paths in the `pieces` directory? One less file to open - a single file is directly linked to every piece location that it contains a delete request for.
4. With the orders file stuff, we ended up with a lot of difficulties when it came to file corruption because we didn't design a file format that could deal with that. We resolved it with a v1 file type (`storagenode/orders/ordersfile/v1.go`) which calculates a checksum for each order, then stores "magic bytes", the checksum, size of the order, and the order. That way we can detect corruption if the checksum we calculate on read differs from the checksum in the file, and if we run into corruption, we can seek to the next, uncorrupted order by looking for the magic bytes. It might be worthwhile to implement a similar process for these files. Definitely don't assume the file will just remain perfectly intact 100% of the time like I did.